qualtran/bloqs/mcmt/multi_control_multi_target_pauli.py

```diff
@@ -141,30 +143,34 @@ def concrete_cvs(self) -> Tuple[int, ...]:
             raise ValueError(f"{self.cvs} is symbolic")
         return self.cvs
 
-    def decompose_from_registers(
-        self, *, context: cirq.DecompositionContext, **quregs: NDArray['cirq.Qid']
-    ) -> Iterator[cirq.OP_TREE]:
-        controls, target = quregs.get('controls', np.array([])), quregs['target']
-        if len(self.concrete_cvs) < 2:
-            controls = controls.flatten()
-            yield [cirq.X(q) for cv, q in zip(self.concrete_cvs, controls) if cv == 0]
-            yield self.target_gate.on(*target).controlled_by(*controls)
-            yield [cirq.X(q) for cv, q in zip(self.concrete_cvs, controls) if cv == 0]
-            return
-        qm = context.qubit_manager
-        and_ancilla, and_target = np.array(qm.qalloc(len(self.concrete_cvs) - 2)), qm.qalloc(1)
-        if len(self.concrete_cvs) == 2:
-            and_op = And(*self.concrete_cvs).on(*controls.flatten(), *and_target)
-            and_op_inv = And(*self.concrete_cvs).adjoint()(*controls.flatten(), *and_target)
-        else:
-            and_op = MultiAnd(self.concrete_cvs).on_registers(
-                ctrl=controls, junk=and_ancilla[:, np.newaxis], target=and_target
-            )
-            and_op_inv = and_op**-1  # type: ignore[operator]
-        yield and_op
-        yield self.target_gate.on(*target).controlled_by(*and_target)
-        yield and_op_inv
-        qm.qfree([*and_ancilla, *and_target])
+    @property
+    def target_bloq(self) -> Bloq:
+        from qualtran.cirq_interop import cirq_gate_to_bloq
+
+        return cirq_gate_to_bloq(self.target_gate)
+
+    @property
+    def _multi_ctrl_bloq(self) -> ControlledViaAnd:
+        ctrl_spec = CtrlSpec(cvs=(np.array(self.concrete_cvs),))
+        return ControlledViaAnd(self.target_bloq, ctrl_spec)
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        if is_symbolic(self.cvs):
+            raise DecomposeTypeError(f"cannot decompose {self} with symbolic {self.cvs=}")
+
+        target = soqs.pop('target')
+        (target_reg_name,) = [reg.name for reg in self.target_bloq.signature]
+        if self.n_ctrls == 0:
+            # TODO discuss if we should remove support for this case.
```

Comment:
This was specifically added after https://github.com/quantumlib/Qualtran/issues/677